gazebo_plugins/src/gazebo_ros_ray_sensor.cpp

```diff
@@ -0,0 +1,428 @@
+      pc.points.push_back(point);
+      pc.channels[0].values.push_back(intensity);
+    }
+  }
+
+  // Publish output
+  boost::get<PointCloudPub>(pub_)->publish(pc);
+}
+
+void GazeboRosRaySensorPrivate::PublishPointCloud2(ConstLaserScanStampedPtr & _msg)
+{
+  // Create message to send
+  sensor_msgs::msg::PointCloud2 pc;
+
+  // Fill header
+  pc.header.frame_id = frame_name_;
+  pc.header.stamp.sec = _msg->time().sec();
+  pc.header.stamp.nanosec = _msg->time().nsec();
+
+  // Create fields in pointcloud
+  sensor_msgs::PointCloud2Modifier pcd_modifier(pc);
+  pcd_modifier.setPointCloud2Fields(4,
+    "x", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "y", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "z", 1, sensor_msgs::msg::PointField::FLOAT32,
+    "intensity", 1, sensor_msgs::msg::PointField::FLOAT32);
+  pcd_modifier.resize(verticalRangeCount * rangeCount);
+  pc.is_dense = true;
+  sensor_msgs::PointCloud2Iterator<float> iter_x(pc, "x");
+  sensor_msgs::PointCloud2Iterator<float> iter_y(pc, "y");
+  sensor_msgs::PointCloud2Iterator<float> iter_z(pc, "z");
+  sensor_msgs::PointCloud2Iterator<float> iter_intensity(pc, "intensity");
+
+  // Fill pointcloud with laser scan
+  for (int i = 0; i < rangeCount; i++) {
+    // Calculate azimuth (horizontal angle)
+    double azimuth;
+    if (rangeCount > 1) {
+      azimuth = i * yDiff / (rangeCount - 1) + minAngle;
+    } else {
+      azimuth = minAngle;
+    }
+    double c_azimuth = cos(azimuth);
+    double s_azimuth = sin(azimuth);
+
+    for (int j = 0; j < verticalRangeCount;
+      ++j, ++iter_x, ++iter_y, ++iter_z, ++iter_intensity)
+    {
+      double inclination;
+      // Calculate inclination (vertical angle)
+      if (verticalRayCount > 1) {
+        inclination = j * pDiff / (verticalRangeCount - 1) + verticalMinAngle;
+      } else {
+        inclination = verticalMinAngle;
+      }
+      double c_inclination = cos(inclination);
+      double s_inclination = sin(inclination);
+
+      // Get range and intensity at this scan
+      size_t index = i + j * rangeCount;
+      double r = _msg->scan().ranges(index);
+      double intensity = _msg->scan().intensities(index);
+      if (intensity < min_intensity_) {
+        intensity = min_intensity_;
+      }
+
+      // Convert spherical coordinates to cartesian for pointcloud
+      // See https://en.wikipedia.org/wiki/Spherical_coordinate_system
+      *iter_x = r * c_inclination * c_azimuth;
+      *iter_y = r * c_inclination * s_azimuth;
+      *iter_z = r * s_inclination;
+      *iter_intensity = intensity;
+    }
+  }
+
+  // Publish output
+  boost::get<PointCloud2Pub>(pub_)->publish(pc);
+}
+
+void GazeboRosRaySensorPrivate::PublishRange(ConstLaserScanStampedPtr & _msg)
+{
+  sensor_msgs::msg::Range range_msg;
+  range_msg.header.frame_id = frame_name_;
+  range_msg.header.stamp.sec = _msg->time().sec();
+  range_msg.header.stamp.nanosec = _msg->time().nsec();
+  range_msg.field_of_view = std::max(pDiff, yDiff);
```

Comment:
On `gazebo_ros_range` the FOV is set from the SDF but now it comes from the sensor itself. If the `<fov>` param wasn't useful, let's document it.